File: JS/BOJ-S4-1620-I_am_Dasom_Pokemon_Master.js
'use strict';

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
let input = fs.readFileSync(filePath).toString().trim().split('\n');

let output = [];
const solution = (input) => {
  const [n, m] = input[0].split(' ').map((v) => +v);

  const mp = new Map();
  for (let i = 1; i <= n; ++i) {
    mp.set(input[i], i);
  }
  for (let i = n + 1; i <= n + m; ++i) {
    if (!Number.isNaN(+input[i])) {
      output.push(input[input[i]]);
    } else {
      output.push(mp.get(input[i]));
    }
  }
};

solution(input);
fs.writeFileSync('/dev/stdout', output.join('\n').trimEnd());
// fs.writeFileSync(
//   '../output.txt',
//   output.join('\n').trimEnd(),
//   (err) => console.error
// );

File: CPP/BOJ-S4-1620-I_am_Dasom_Pokemon_Master.cpp
#include <bits/stdc++.h>
using namespace std;

// int n, m;
// string s;
// map<string, int> mp;
// map<int, string> mp2;
// string a[100004];

// int main()
// {
//   ios_base::sync_with_stdio(false);
//   cin.tie(NULL);
//   cout.tie(NULL);
//   // freopen("input.txt", "r", stdin);
//   // freopen("output.txt", "w", stdout);

//   cin >> n >> m;
//   for (int i = 0; i < n; ++i)
//   {
//     cin >> s;
//     mp[s] = i + 1;
//     // mp2[i + 1] = s;
//     a[i + 1] = s;
//   }

//   for (int i = 0; i < m; ++i)
//   {
//     cin >> s;
//     if (atoi(s.c_str()) == 0)
//     {
//       cout << mp[s] << '\n';
//     }
//     else
//     {
//       cout << a[atoi(s.c_str())] << '\n';
//       // cout << mp2[atoi(s.c_str())] << '\n';
//     }
//   }

//   return 0;
// }

int n, m;
unordered_map<string, int> mp;
string s, a[100004];

int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  freopen("input.txt", "r", stdin);
  freopen("output.txt", "w", stdout);

  cin >> n >> m;
  for (int i = 0; i < n; ++i)
  {
    cin >> s;
    a[i + 1] = s;
    mp[s] = i + 1;
  }

  for (int i = 0; i < m; ++i)
  {
    cin >> s;
    if (atoi(s.c_str()) == 0)
    {
      cout << mp[s] << '\n';
    }
    else
    {
      cout << a[atoi(s.c_str())] << '\n';
    }
  }

  return 0;
}